skbio/stats/distance/_base.py

```diff
@@ -738,6 +740,118 @@ def from_iterable(cls, iterable, metric, key=None, keys=None):
 
         return cls(dm, keys_)
 
+    @classonlymethod
+    @experimental(as_of="0.4.2-dev")
+    def from_pairwise_distances(cls, pairwise_distances):
+        """Create ``DistanceMatrix`` from an iterable of ``PairwiseDistances``.
+
+        Constructs a ``DistanceMatrix`` from the complete set of IDs defined by
+        `pairwise_distances`. All "between" pairwise distances must be defined.
+
+        Parameters
+        ----------
+        pairwise_distances : iterable of PairwiseDistances
+            Pairwise distances to merge into a single ``DistanceMatrix``.
+
+        Returns
+        -------
+        DistanceMatrix
+            Distance matrix representing all pairwise distances between the
+            complete set of IDs present in `pairwise_distances`. The IDs in the
+            distance matrix will be in sorted order as there is no ordering
+            defined by `pairwise_distances`.
+
+        Raises
+        ------
+        TypeError
+            If `pairwise_distances` does not contain ``PairwiseDistances``
+            objects.
+        ValueError
+            If there are conflicting distances between a pair of IDs (e.g., two
+            ``PairwiseDistances`` objects define a different distance between
+            the same pair of IDs).
+        ValueError
+            If there is not at least one ID defined across
+            `pairwise_distances`.
+        ValueError
+            If any "between" distances are missing.
+
+        See Also
+        --------
+        skbio.stats.distance.PairwiseDistances
+        skbio.diversity.beta_diversity
+
+        Examples
+        --------
+        Construct a ``DistanceMatrix`` from two ``PairwiseDistances`` objects
+        collectively defining three IDs, `'A'`, `'B'`, and `'C'`:
+
+        >>> from skbio.stats.distance import DistanceMatrix, PairwiseDistances
+        >>> pwdists = [
+        ...     PairwiseDistances({('A', 'B'): 0.3, ('A', 'C'): 0.1}),
+        ...     PairwiseDistances({('B', 'C'): 0.2})
+        ... ]
+        >>> dm = DistanceMatrix.from_pairwise_distances(pwdists)
+        >>> dm
+        3x3 distance matrix
+        IDs:
+        'A', 'B', 'C'
+        Data:
+        [[ 0.   0.3  0.1]
+         [ 0.3  0.   0.2]
+         [ 0.1  0.2  0. ]]
+
+        Note that the IDs are in sorted order.
+
+        """
+        distances = {}
```

Comment:
There is actually a huge difference. I did some tests to actually check the performance of the solution presented here and the solution proposed by @wasade .

`chocker.py`:
```python
    d = {}
    for i in range(num_samples):
        for j in range(i+1, num_samples):
            print i, j
            if (i, j) in d or (j, i) in d:
                continue
            else:
                d[(i, j)] = 1

    print d
```

`no_chocker_numpy.py`
```python
    d = np.zeros([num_samples, num_samples])
    for i in range(num_samples):
        for j in range(i+1, num_samples):
            print i, j
            if (i, j) in d or (j, i) in d:
                continue
            else:
                d[i][j] = 1.0

    print d
```

I tried with `num_samples = 10000` and `num_samples = 20000`.

| solution    | memory (GB) | time  |
|-------------|-------------|-------|
| chocker 10k | ~34GB       | 4:32  |
| numpy 10k   | ~3GB        | 8:00  |
| chocker 20k | ~137GB      | 56:00 |
| numpy 20k   | ~12GB       | 39:04 |

@wasade this will explain why your laptop was really unhappy when you tried with 25k hahaha

As we can see, as we scale up, we no only use less memory, we are also faster. In this case, storing a dense representation is more efficient than relying on python dicts.

The memory values above are the ones reported by maxresident memory in /usr/bin/time and the time is elapsed time.